File: converted_source/upp_int.pas.py
# unit upp_int

from conversion_common import *
import udebug
import udomain
import usupport
import uunits
import umath
import updcom
import umain
import usliders
import uppanel
import uparams
import uplant
import delphi_compatability

# const
kItemValueText = 1
kItemSlider = 2

#  kItemUnitText = 2;    
class PdSmallintParameterPanel(PdParameterPanel):
    def __init__(self):
        self.currentValue = 0
        self.currentValueWhileSliding = 0
        self.minValue = 0
        self.maxValue = 0
        self.slider = KfSlider()
        self.unitIndex = 0
        self.minTextRect = TRect()
        self.maxTextRect = TRect()
        self.valueTextRect = TRect()
        self.unitTextRect = TRect()
        self.drawWithoutBounds = false
    
    def initialize(self):
        #assuming slider will be deleted automatically by owner - self - otherwise would have destructor
        self.slider = usliders.KfSlider().create(self)
        self.slider.Parent = self
        self.slider.FOnMouseDown = self.sliderMouseDown
        self.slider.FOnMouseMove = self.sliderMouseMove
        self.slider.FOnMouseUp = self.sliderMouseUp
        self.slider.FOnKeyDown = self.sliderKeyDown
        self.slider.readOnly = self.param.readOnly
        self.slider.useDefaultSizeAndDraggerSize()
        self.minValue = intround(self.param.lowerBound())
        self.maxValue = intround(self.param.upperBound())
        self.slider.minValue = self.minValue
        self.slider.maxValue = self.maxValue
    
    def checkValueAgainstBounds(self, value):
        result = 0
        result = value
        if result < self.minValue:
            result = self.minValue
        if result > self.maxValue:
            result = self.maxValue
        return result
    
    def doMouseUp(self, Sender, Button, Shift, X, Y):
        thePoint = TPoint()
        
        # must always call this first because it sets the focus 
        PdParameterPanel.doMouseUp(self, Sender, Button, Shift, X, Y)
        thePoint = Point(X, Y)
        if delphi_compatability.PtInRect(self.valueTextRect, thePoint):
            if self.slider.Enabled and not self.slider.readOnly:
                self.selectedItemIndex = kItemValueText
                self.Invalidate()
                self.adjustValue()
    
    def doKeyDown(self, sender, key, shift):
        if self.slider.Enabled:
            if key == delphi_compatability.VK_HOME:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
            elif key == delphi_compatability.VK_END:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
            elif key == delphi_compatability.VK_DOWN:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
            elif key == delphi_compatability.VK_LEFT:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
            elif key == delphi_compatability.VK_UP:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
            elif key == delphi_compatability.VK_RIGHT:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
            elif key == delphi_compatability.VK_NEXT:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
            elif key == delphi_compatability.VK_PRIOR:
                if (self.selectedItemIndex == kItemSlider):
                    # process slider arrow keys first 
                    key = self.slider.doKeyDown(sender, key, shift)
                    return key
        if (self.selectedItemIndex == kItemValueText) and (key == delphi_compatability.VK_RETURN):
            key = 0
            self.Invalidate()
            self.adjustValue()
            return key
        PdParameterPanel.doKeyDown(self, sender, key, shift)
        if self.slider.Enabled:
            if (key == delphi_compatability.VK_RETURN) and (self.selectedItemIndex == kItemSlider):
                key = self.slider.doKeyDown(sender, key, shift)
        return key
    
    def adjustValue(self):
        newString = ansistring()
        oldValue = 0
        newValue = 0
        oldString = ""
        prompt = ""
        nameString = ""
        unitString = ""
        
        oldValue = self.currentValue
        newString = IntToStr(self.currentValue)
        oldString = newString
        nameString = UNRESOLVED.copy(self.Caption, 1, 30)
        if len(nameString) == 30:
            nameString = nameString + "..."
        unitString = uunits.UnitStringForEnum(self.param.unitSet, self.unitIndex)
        prompt = "Type a new value."
        if InputQuery("Enter value", prompt, newString):
            if (newString != oldString):
                newValue = StrToIntDef(newString, 0)
                if self.checkValueAgainstBounds(newValue) == newValue:
                    self.currentValue = newValue
                    umain.MainForm.doCommand(updcom.PdChangeSmallintValueCommand().createCommandWithListOfPlants(umain.MainForm.selectedPlants, self.currentValue, self.param.fieldNumber, uparams.kNotArray, self.param.regrow))
                    self.updateDisplay()
                    self.Invalidate()
    
    def updateEnabling(self):
        PdParameterPanel.updateEnabling(self)
        self.slider.Enabled = self.plant != None
        self.slider.readOnly = self.readOnly
    
    def updatePlantValues(self):
        oldValue = 0
        
        if self.plant == None:
            return
        if self.param.collapsed:
            return
        oldValue = self.currentValue
        self.updateCurrentValue(-1)
        if oldValue != self.currentValue:
            self.updateDisplay()
            self.Invalidate()
    
    def updateCurrentValue(self, aFieldIndex):
        if (self.plant != None) and (self.param.fieldType == uparams.kFieldSmallint):
            self.currentValue = self.plant.transferField(umath.kGetField, self.currentValue, self.param.fieldNumber, uparams.kFieldSmallint, uparams.kNotArray, false, None)
            if self.currentValue < self.minValue:
                self.minValue = self.currentValue
                self.slider.minValue = self.minValue
            if self.currentValue > self.maxValue:
                self.maxValue = self.currentValue
                self.slider.maxValue = self.maxValue
        else:
            self.currentValue = 0
    
    def updateDisplay(self):
        if (self.plant != None) and (self.param.fieldType == uparams.kFieldSmallint):
            self.slider.currentValue = self.currentValue
        else:
            self.slider.currentValue = 0
            self.slider.Enabled = false
    
    def sliderMouseDown(self, Sender):
        if self.plant == None:
            return
        self.selectedItemIndex = kItemSlider
        self.slider.hasUnofficialFocus = true
        self.currentValueWhileSliding = self.slider.currentValue
        self.Invalidate()
        if not self.Focused():
            self.SetFocus()
    
    def sliderMouseMove(self, Sender):
        if self.plant == None:
            return
        self.currentValueWhileSliding = self.slider.currentValue
        self.Invalidate()
    
    def sliderMouseUp(self, Sender):
        if self.plant != None:
            self.currentValue = self.slider.currentValue
            umain.MainForm.doCommand(updcom.PdChangeSmallintValueCommand().createCommandWithListOfPlants(umain.MainForm.selectedPlants, self.currentValue, self.param.fieldNumber, uparams.kNotArray, self.param.regrow))
            self.updateDisplay()
            self.Invalidate()
        else:
            self.slider.currentValue = 0
            self.slider.Enabled = false
    
    def sliderKeyDown(self, sender):
        self.sliderMouseUp(sender)
    
    def maxWidth(self):
        result = 0
        result = uppanel.kLeftRightGap * 2 + umath.intMax(self.labelWidth, self.minScaleWidthWithBounds())
        return result
    
    def minWidth(self, requestedWidth):
        result = 0
        widthForLongestWord = 0
        widthWithBounds = 0
        widthWithoutBounds = 0
        
        widthForLongestWord = uppanel.kLeftRightGap * 2 + self.longestLabelWordWidth
        widthWithBounds = self.minScaleWidthWithBounds()
        widthWithoutBounds = self.minScaleWidthWithoutBounds()
        result = -1
        if requestedWidth < widthForLongestWord:
            result = widthForLongestWord
        if requestedWidth < widthWithBounds:
            self.drawWithoutBounds = true
            if requestedWidth < widthWithoutBounds:
                result = umath.intMax(result, widthWithoutBounds)
        else:
            self.drawWithoutBounds = false
        return result
    
    def minScaleWidthWithBounds(self):
        result = 0
        minString = ""
        valueString = ""
        unitString = ""
        maxString = ""
        
        minString = IntToStr(self.minValue)
        valueString = IntToStr(self.currentValue) + "  "
        unitString = uunits.UnitStringForEnum(self.param.unitSet, self.unitIndex)
        maxString = IntToStr(self.maxValue)
        result = self.formTextWidth(minString + valueString + unitString + maxString) + udebug.kBetweenGap * 3
        return result
    
    def minScaleWidthWithoutBounds(self):
        result = 0
        valueString = ""
        unitString = ""
        
        valueString = IntToStr(self.currentValue) + "  "
        unitString = uunits.UnitStringForEnum(self.param.unitSet, self.unitIndex)
        result = self.formTextWidth(valueString + unitString) + udebug.kBetweenGap
        return result
    
    def uncollapsedHeight(self):
        result = 0
        result = self.collapsedHeight() + self.textHeight + uppanel.kTopBottomGap
        return result
    
    def resizeElements(self):
        self.paint()
        self.slider.Left = self.minTextRect.Right + uppanel.kSliderSpaceGap
        self.slider.Top = self.collapsedHeight() + self.textHeight / 2 - self.slider.Height / 2
        self.slider.Width = umath.intMax(0, self.maxTextRect.Left - self.minTextRect.Right - uppanel.kSliderSpaceGap * 2)
    
    def maxSelectedItemIndex(self):
        result = 0
        if (not self.param.collapsed) and (self.slider.Enabled):
            result = kItemSlider
        else:
            result = uppanel.kItemLabel
        return result
    
    def paint(self):
        Rect = TRect()
        minString = ""
        valueString = ""
        maxString = ""
        unitString = ""
        
        if delphi_compatability.Application.terminated:
            return
        if (self.selectedItemIndex == kItemSlider):
            # ask slider to update itself based on whether it is selected 
            self.slider.hasUnofficialFocus = true
        else:
            self.slider.hasUnofficialFocus = false
        PdParameterPanel.paint(self)
        self.slider.Visible = not self.param.collapsed
        if (self.param.collapsed):
            self.drawExtraValueCaptionWithString(IntToStr(self.currentValue), uunits.UnitStringForEnum(self.param.unitSet, self.unitIndex))
            return
        Rect = self.GetClientRect()
        minString = IntToStr(self.minValue)
        if self.slider.dragging:
            valueString = IntToStr(self.currentValueWhileSliding)
        else:
            valueString = IntToStr(self.currentValue)
        unitString = uunits.UnitStringForEnum(self.param.unitSet, self.unitIndex)
        maxString = IntToStr(self.maxValue)
        self.valueTextRect.Left = uppanel.kLeftRightGap
        self.valueTextRect.Right = self.valueTextRect.Left + self.Canvas.TextWidth(valueString)
        self.valueTextRect.Top = self.collapsedHeight() + 1
        self.valueTextRect.Bottom = self.valueTextRect.Top + self.textHeight
        self.unitTextRect = self.valueTextRect
        self.unitTextRect.Left = self.valueTextRect.Right + 5
        self.unitTextRect.Right = self.unitTextRect.Left + self.Canvas.TextWidth(unitString)
        self.minTextRect = self.valueTextRect
        self.minTextRect.Left = self.unitTextRect.Right + 5
        self.minTextRect.Right = self.minTextRect.Left + self.Canvas.TextWidth(minString)
        self.maxTextRect = self.valueTextRect
        self.maxTextRect.Right = Rect.right - uppanel.kLeftRightGap
        self.maxTextRect.Left = self.maxTextRect.Right - self.Canvas.TextWidth(maxString)
        self.drawText(minString, self.minTextRect, false, false, true)
        self.drawText(valueString, self.valueTextRect, not self.readOnly, (self.selectedItemIndex == kItemValueText), false)
        self.drawText(unitString, self.unitTextRect, false, false, false)
        self.drawText(maxString, self.maxTextRect, false, false, true)
    

File: converted_source/upp_int.pas.java
// unit upp_int

from conversion_common import *;
import udebug;
import usliders;
import uppanel;
import delphi_compatability;

// const
kItemValueText = 1;
kItemSlider = 2;



//  kItemUnitText = 2;    
class PdSmallintParameterPanel extends PdParameterPanel {
    public short currentValue;
    public short currentValueWhileSliding;
    public short minValue;
    public short maxValue;
    public KfSlider slider;
    public short unitIndex;
    public TRect minTextRect;
    public TRect maxTextRect;
    public TRect valueTextRect;
    public TRect unitTextRect;
    public boolean drawWithoutBounds;
    
    public void initialize() {
        //assuming slider will be deleted automatically by owner - self - otherwise would have destructor
        this.slider = usliders.KfSlider().create(this);
        this.slider.Parent = this;
        this.slider.FOnMouseDown = this.sliderMouseDown;
        this.slider.FOnMouseMove = this.sliderMouseMove;
        this.slider.FOnMouseUp = this.sliderMouseUp;
        this.slider.FOnKeyDown = this.sliderKeyDown;
        this.slider.readOnly = this.param.readOnly;
        this.slider.useDefaultSizeAndDraggerSize();
        this.minValue = intround(this.param.lowerBound);
        this.maxValue = intround(this.param.upperBound);
        this.slider.minValue = this.minValue;
        this.slider.maxValue = this.maxValue;
    }
    
    public short checkValueAgainstBounds(short value) {
        result = 0;
        result = value;
        if (result < this.minValue) {
            result = this.minValue;
        }
        if (result > this.maxValue) {
            result = this.maxValue;
        }
        return result;
    }
    
    public void doMouseUp(TObject Sender, TMouseButton Button, TShiftState Shift, int X, int Y) {
        TPoint thePoint = new TPoint();
        
        // must always call this first because it sets the focus 
        super.doMouseUp(Sender, Button, Shift, X, Y);
        thePoint = Point(X, Y);
        if (delphi_compatability.PtInRect(this.valueTextRect, thePoint)) {
            if (this.slider.Enabled && !this.slider.readOnly) {
                this.selectedItemIndex = kItemValueText;
                this.Invalidate();
                this.adjustValue();
            }
        }
    }
    
    public void doKeyDown(TObject sender, byte key, TShiftState shift) {
        if (this.slider.Enabled) {
            switch (key) {
                case delphi_compatability.VK_HOME:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
                case delphi_compatability.VK_END:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
                case delphi_compatability.VK_DOWN:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
                case delphi_compatability.VK_LEFT:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
                case delphi_compatability.VK_UP:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
                case delphi_compatability.VK_RIGHT:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
                case delphi_compatability.VK_NEXT:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
                case delphi_compatability.VK_PRIOR:
                    if ((this.selectedItemIndex == kItemSlider)) {
                        // process slider arrow keys first 
                        key = this.slider.doKeyDown(sender, key, shift);
                        return key;
                    }
                    break;
        }
        if ((this.selectedItemIndex == kItemValueText) && (key == delphi_compatability.VK_RETURN)) {
            key = 0;
            this.Invalidate();
            this.adjustValue();
            return key;
        }
        super.doKeyDown(sender, key, shift);
        if (this.slider.Enabled) {
            if ((key == delphi_compatability.VK_RETURN) && (this.selectedItemIndex == kItemSlider)) {
                key = this.slider.doKeyDown(sender, key, shift);
            }
        }
        return key;
    }
    
    public void adjustValue() {
        ansistring newString = new ansistring();
        short oldValue = 0;
        short newValue = 0;
        String oldString = "";
        String prompt = "";
        String nameString = "";
        String unitString = "";
        
        oldValue = this.currentValue;
        newString = IntToStr(this.currentValue);
        oldString = newString;
        nameString = UNRESOLVED.copy(this.Caption, 1, 30);
        if (len(nameString) == 30) {
            nameString = nameString + "...";
        }
        unitString = UNRESOLVED.UnitStringForEnum(this.param.unitSet, this.unitIndex);
        prompt = "Type a new value.";
        if (InputQuery("Enter value", prompt, newString)) {
            if ((newString != oldString)) {
                newValue = StrToIntDef(newString, 0);
                if (this.checkValueAgainstBounds(newValue) == newValue) {
                    this.currentValue = newValue;
                    UNRESOLVED.MainForm.doCommand(UNRESOLVED.PdChangeSmallintValueCommand.createCommandWithListOfPlants(UNRESOLVED.MainForm.selectedPlants, this.currentValue, this.param.fieldNumber, UNRESOLVED.kNotArray, this.param.regrow));
                    this.updateDisplay();
                    this.Invalidate();
                }
            }
        }
    }
    
    public void updateEnabling() {
        super.updateEnabling();
        this.slider.Enabled = this.plant != null;
        this.slider.readOnly = this.readOnly;
    }
    
    public void updatePlantValues() {
        int oldValue = 0;
        
        if (this.plant == null) {
            return;
        }
        if (this.param.collapsed) {
            return;
        }
        oldValue = this.currentValue;
        this.updateCurrentValue(-1);
        if (oldValue != this.currentValue) {
            this.updateDisplay();
            this.Invalidate();
        }
    }
    
    public void updateCurrentValue(int aFieldIndex) {
        if ((this.plant != null) && (this.param.fieldType == UNRESOLVED.kFieldSmallint)) {
            this.plant.transferField(UNRESOLVED.kGetField, this.currentValue, this.param.fieldNumber, UNRESOLVED.kFieldSmallint, UNRESOLVED.kNotArray, false, null);
            if (this.currentValue < this.minValue) {
                this.minValue = this.currentValue;
                this.slider.minValue = this.minValue;
            }
            if (this.currentValue > this.maxValue) {
                this.maxValue = this.currentValue;
                this.slider.maxValue = this.maxValue;
            }
        } else {
            this.currentValue = 0;
        }
    }
    
    public void updateDisplay() {
        if ((this.plant != null) && (this.param.fieldType == UNRESOLVED.kFieldSmallint)) {
            this.slider.currentValue = this.currentValue;
        } else {
            this.slider.currentValue = 0;
            this.slider.Enabled = false;
        }
    }
    
    public void sliderMouseDown(TObject Sender) {
        if (this.plant == null) {
            return;
        }
        this.selectedItemIndex = kItemSlider;
        this.slider.hasUnofficialFocus = true;
        this.currentValueWhileSliding = this.slider.currentValue;
        this.Invalidate();
        if (!this.Focused()) {
            this.SetFocus();
        }
    }
    
    public void sliderMouseMove(TObject Sender) {
        if (this.plant == null) {
            return;
        }
        this.currentValueWhileSliding = this.slider.currentValue;
        this.Invalidate();
    }
    
    public void sliderMouseUp(TObject Sender) {
        if (this.plant != null) {
            this.currentValue = this.slider.currentValue;
            UNRESOLVED.MainForm.doCommand(UNRESOLVED.PdChangeSmallintValueCommand.createCommandWithListOfPlants(UNRESOLVED.MainForm.selectedPlants, this.currentValue, this.param.fieldNumber, UNRESOLVED.kNotArray, this.param.regrow));
            this.updateDisplay();
            this.Invalidate();
        } else {
            this.slider.currentValue = 0;
            this.slider.Enabled = false;
        }
    }
    
    public void sliderKeyDown(TOBject sender) {
        this.sliderMouseUp(sender);
    }
    
    public int maxWidth() {
        result = 0;
        result = uppanel.kLeftRightGap * 2 + UNRESOLVED.intMax(this.labelWidth, this.minScaleWidthWithBounds());
        return result;
    }
    
    public int minWidth(int requestedWidth) {
        result = 0;
        int widthForLongestWord = 0;
        int widthWithBounds = 0;
        int widthWithoutBounds = 0;
        
        widthForLongestWord = uppanel.kLeftRightGap * 2 + this.longestLabelWordWidth;
        widthWithBounds = this.minScaleWidthWithBounds();
        widthWithoutBounds = this.minScaleWidthWithoutBounds();
        result = -1;
        if (requestedWidth < widthForLongestWord) {
            result = widthForLongestWord;
        }
        if (requestedWidth < widthWithBounds) {
            this.drawWithoutBounds = true;
            if (requestedWidth < widthWithoutBounds) {
                result = UNRESOLVED.intMax(result, widthWithoutBounds);
            }
        } else {
            this.drawWithoutBounds = false;
        }
        return result;
    }
    
    public int minScaleWidthWithBounds() {
        result = 0;
        String minString = "";
        String valueString = "";
        String unitString = "";
        String maxString = "";
        
        minString = IntToStr(this.minValue);
        valueString = IntToStr(this.currentValue) + "  ";
        unitString = UNRESOLVED.UnitStringForEnum(this.param.unitSet, this.unitIndex);
        maxString = IntToStr(this.maxValue);
        result = this.formTextWidth(minString + valueString + unitString + maxString) + udebug.kBetweenGap * 3;
        return result;
    }
    
    public int minScaleWidthWithoutBounds() {
        result = 0;
        String valueString = "";
        String unitString = "";
        
        valueString = IntToStr(this.currentValue) + "  ";
        unitString = UNRESOLVED.UnitStringForEnum(this.param.unitSet, this.unitIndex);
        result = this.formTextWidth(valueString + unitString) + udebug.kBetweenGap;
        return result;
    }
    
    public int uncollapsedHeight() {
        result = 0;
        result = this.collapsedHeight() + this.textHeight + uppanel.kTopBottomGap;
        return result;
    }
    
    public void resizeElements() {
        this.paint();
        this.slider.Left = this.minTextRect.Right + uppanel.kSliderSpaceGap;
        this.slider.Top = this.collapsedHeight() + this.textHeight / 2 - this.slider.Height / 2;
        this.slider.Width = UNRESOLVED.intMax(0, this.maxTextRect.Left - this.minTextRect.Right - uppanel.kSliderSpaceGap * 2);
    }
    
    public int maxSelectedItemIndex() {
        result = 0;
        if ((!this.param.collapsed) && (this.slider.Enabled)) {
            result = kItemSlider;
        } else {
            result = uppanel.kItemLabel;
        }
        return result;
    }
    
    public void paint() {
        TRect Rect = new TRect();
        String minString = "";
        String valueString = "";
        String maxString = "";
        String unitString = "";
        
        if (delphi_compatability.Application.terminated) {
            return;
        }
        if ((this.selectedItemIndex == kItemSlider)) {
            // ask slider to update itself based on whether it is selected 
            this.slider.hasUnofficialFocus = true;
        } else {
            this.slider.hasUnofficialFocus = false;
        }
        super.paint();
        this.slider.Visible = !this.param.collapsed;
        if ((this.param.collapsed)) {
            this.drawExtraValueCaptionWithString(IntToStr(this.currentValue), UNRESOLVED.UnitStringForEnum(this.param.unitSet, this.unitIndex));
            return;
        }
        Rect = this.GetClientRect();
        minString = IntToStr(this.minValue);
        if (this.slider.dragging) {
            valueString = IntToStr(this.currentValueWhileSliding);
        } else {
            valueString = IntToStr(this.currentValue);
        }
        unitString = UNRESOLVED.UnitStringForEnum(this.param.unitSet, this.unitIndex);
        maxString = IntToStr(this.maxValue);
        this.valueTextRect.Left = uppanel.kLeftRightGap;
        this.valueTextRect.Right = this.valueTextRect.Left + this.Canvas.TextWidth(valueString);
        this.valueTextRect.Top = this.collapsedHeight() + 1;
        this.valueTextRect.Bottom = this.valueTextRect.Top + this.textHeight;
        this.unitTextRect = this.valueTextRect;
        this.unitTextRect.Left = this.valueTextRect.Right + 5;
        this.unitTextRect.Right = this.unitTextRect.Left + this.Canvas.TextWidth(unitString);
        this.minTextRect = this.valueTextRect;
        this.minTextRect.Left = this.unitTextRect.Right + 5;
        this.minTextRect.Right = this.minTextRect.Left + this.Canvas.TextWidth(minString);
        this.maxTextRect = this.valueTextRect;
        this.maxTextRect.Right = Rect.right - uppanel.kLeftRightGap;
        this.maxTextRect.Left = this.maxTextRect.Right - this.Canvas.TextWidth(maxString);
        this.drawText(minString, this.minTextRect, false, false, true);
        this.drawText(valueString, this.valueTextRect, !this.readOnly, (this.selectedItemIndex == kItemValueText), false);
        this.drawText(unitString, this.unitTextRect, false, false, false);
        this.drawText(maxString, this.maxTextRect, false, false, true);
    }
    
}
